parsely/src/main/java/com/parsely/parselyandroid/ParselyTracker.kt

```diff
@@ -0,0 +1,99 @@
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+package com.parsely.parselyandroid
+
+import android.content.Context
+
+/**
+ * Tracks Parse.ly app views in Android apps
+ *
+ *
+ * Accessed as a singleton. Maintains a queue of pageview events in memory and periodically
+ * flushes the queue to the Parse.ly pixel proxy server.
+ */
+public interface ParselyTracker {
+    public val engagementInterval: Double?
+    public val videoEngagementInterval: Double?
+    public val flushInterval: Long
+
+    public fun engagementIsActive(): Boolean
+    public fun videoIsActive(): Boolean
+
+    public fun trackPageview(
+        url: String,
+        urlRef: String = "",
+        urlMetadata: ParselyMetadata? = null,
+        extraData: Map<String, Any>? = null,
+    )
+
+    public fun startEngagement(
+        url: String,
+        urlRef: String = "",
+        extraData: Map<String, Any>? = null
+    )
+
+    public fun stopEngagement()
+
+    public fun trackPlay(
+        url: String,
+        urlRef: String = "",
+        videoMetadata: ParselyVideoMetadata,
+        extraData: Map<String, Any>? = null,
+    )
+    public fun trackPause()
+    public fun resetVideo()
+    public fun flushEventQueue()
+    public fun flushTimerIsActive(): Boolean
+
+    public companion object {
+        private const val DEFAULT_FLUSH_INTERVAL_SECS = 60
+        private var instance: ParselyTrackerInternal? = null
+
+        /**
+         * Singleton instance accessor. Note: This must be called after [.sharedInstance]
+         *
+         * @return The singleton instance
+         */
+        @JvmStatic
+        public fun sharedInstance(): ParselyTracker? {
+            return instance
+        }
+
+        /**
+         * Singleton instance factory Note: this must be called before [.sharedInstance]
+         *
+         * @param siteId        The Parsely public site id (eg "example.com")
+         * @param flushInterval The interval at which the events queue should flush, in seconds
+         * @param context             The current Android application context
+         * @param dryRun If set to `true`, events **won't** be sent to Parse.ly server
+         * @return The singleton instance
+         */
+        @JvmStatic
+        @JvmOverloads
+        public fun sharedInstance(
+            siteId: String,
+            flushInterval: Int = DEFAULT_FLUSH_INTERVAL_SECS,
+            context: Context,
+            dryRun: Boolean = false,
+        ): ParselyTracker {
```

Comment:
A singleton that requires a configuration presents an interesting problem. What happens if the client tries to pass a different configuration after the instance is already initialized?

Currently, the secondary configuration is silently ignored. And, I could see a case where this can result in developers trying to figure out why their configuration change is not applied. Assuming that we don't want to allow updating the configuration after the initialization, I think we can give them a heads up about this in the documentation and by printing a log.

How do you feel about:
1. Adding a note to the documentation of this method, saying that this method should only be called once and subsequent calls will ignore its arguments and return the existing instance
2. Either add a generic warning about this ^ as a log every time we return the existing instance. Or - add a specific warning if this method is called with a different configuration as compared to the existing instance.

Note that an alternative approach that would avoid this whole issue would be to use a build time configuration instead of function parameters, but I don't think that's appropriate for this use case. Even if it was, that approach has its own issues - but I wanted to mention it for completeness-sake.